File: Exp 1/CRC_Pro.cpp
// ConsoleApplication1.cpp : 定义控制台应用程序的入口点。
//

#include "stdafx.h"
#include<iostream>
#include<string>
#include<fstream>
#include<bitset>
#include<ctime>
using namespace std;

long long int f(string s)//translate a binary string to long
{
	long long int a = 0;
	for (int i = 0; i < s.length(); i++)
	{
		a = a * 2 + (s[i] - '0');
	}
	return a;
}

string t(long long int a, int length)//translate a long to binary 
{
	string s;
	while (length)
	{
		string temp = "1";
		temp[0] = a % 2 + '0';
		s = temp + s;
		a /= 2;
		length--;
	}
	return s;
}

string tobyte(string mem)
{
	string res;
	for (int i = 0; i < mem.length(); i++)
	{
		res += t(mem[i], 8);
	}
	return res;
}

long long GetRemainder(string newString, string GenXString) {
	int r = GenXString.length() - 1;
	string subStr = newString.substr(0, r);
	long long mod = f(subStr);
	long long divisor = f(GenXString);
	for (int i = r; i < newString.length(); i++) {
		mod = (mod << 1) + newString[i] - 48;
		if (mod & 0x10000) {
			mod = mod ^ divisor;
		}
		else {
			mod = mod ^ 0;
		}
	}
	return mod;
}

string GetSendString(string InfoString1, string GenXString) //形成CRC
{
	string newString = InfoString1;
	for (size_t i = 0; i < GenXString.length() - 1; i++)
	{
		newString += "0";
	}
	long long mod = GetRemainder(newString, GenXString);
	string sendString = InfoString1 + t(mod, 16);
	cout << sendString << endl;
	return sendString;
}

bool isCRC(string InfoString2, string GenXString)//验证CRC正确
{
	long long mod = GetRemainder(InfoString2, GenXString);
	if (mod == 0) {
		return true;
	}
	else {
		return false;
	}
}

int main()
{
	string InfoString1, GenXString;
	const int CRCLength = 16;//数值为GenXString.length()-1;
	InfoString1 = "Hello World";
	GenXString = "10001000000100001";
	cout << InfoString1 << endl;
	string newString = tobyte(InfoString1);
	string SendString = GetSendString(newString, GenXString);
	cout << "Send " << SendString << endl;
	cout <<"Receive "<< SendString << endl;
	if (isCRC(SendString, GenXString))
	{
		cout << "Success" << endl;
	}
	else
	{
		cout << "Error" << endl;
	}
	system("pause");
}

File: Exp 1/CRC_Pro.py
#!/usr/bin/python3
import math

def GetRemainder(newString, GenXString):
    r = len(GenXString) - 1
    SubString = newString[0:r]
    mod = int(SubString, 2)
    divisor = int(GenXString, 2)
    for i in range(r,len(newString)):
        mod = mod * 2 + int(newString[i], 2)
        if mod >= math.pow(2, r):
            mod ^= divisor
        else:
            mod ^=0
    return mod

def GetSendString(InfoString1, GenXString):
    rr = len(GenXString) - 1
    newString = InfoString1
    for i in range(0, rr):
        newString = newString + "0"
    mod = GetRemainder(newString, GenXString)
    print("CRC-Code = ",bin(mod).replace("0b",""))
    tmp = int(newString, 2)
    return bin(mod ^ tmp).replace("0b","")


print("Sender:")
InfoString1 = "11011010110010010100010110"
GenXString = "10001000000100001"    #17位
print("InfoString1 = ",InfoString1)
print("GenXString = ",GenXString)
SendString = GetSendString(InfoString1, GenXString)
print("SendString = ", SendString)
print("\n")
print("Receiver:")
InfoString2 = SendString
#InfoString2 = SendString[:-1] + str(int(SendString[-1]) ^ 1)  #最后一位取反
cmp = GetRemainder(InfoString2, GenXString)
print("InfoString2 = ",InfoString2)
print("CRC-Code = ", bin(cmp).replace("0b",""))
if cmp == 0:
    print("Correct result.")
else:
    print("Wrong result.")
input("Please enter...")
